src/dtree.ml

```diff
@@ -0,0 +1,499 @@
+              Seq.map (fun (te, po) -> (te, Subterm.prefix p po)) |>
+              Array.of_seq
+      | _             -> assert false
+
+  (** [specialize p c m] specializes the matrix [m] when matching pattern [p]
+      against column [c].  A matrix can be specialized by a user defined
+      symbol.  In case an {!constructor:Appl} is given as pattern [p], only
+      terms having the same number of arguments and the same leftmost {e non}
+      {!constructor:Appl} term match. *)
+  let specialize : term -> int -> rule list -> rule list = fun p ci rs ->
+    let filtered = List.filter (fun { lhs ; _} ->
+      spec_filter p (fst lhs.(ci))) rs in
+    let newcith = List.map (fun { lhs ; _ } -> spec_transform p lhs.(ci))
+      filtered in
+    List.map2 (fun newcith rul ->
+      let { lhs ; _ } = rul in
+      let postfix = Array.drop (ci + 1) lhs in
+      let newline = Array.concat
+        [ Array.sub lhs 0 ci
+        ; newcith
+        ; postfix ] in
+      { rul with lhs = newline }) newcith filtered
+
+  (** [default c m] computes the default matrix containing what remains to be
+      matched in case the pattern used is not in the column [c]. *)
+  let default : int -> rule list -> rule list = fun ci rs ->
+    let filtered = List.filter (fun { lhs ; _ } ->
+      match fst @@ lhs.(ci) with
+      | Patt(_ , _, _) -> true
+      | Symb(_, _)
+      | Abst(_, _)
+      | Vari(_)
+      | Appl(_, _)     -> false
+      | _              -> assert false) rs in
+    List.map (fun rul ->
+      let { lhs ; _ } = rul in
+      match lhs.(ci) with
+      | Patt(_, _, _), _ ->
+          let postfix = Array.drop (ci + 1) lhs in
+          { rul with lhs = Array.append (Array.sub lhs 0 ci) postfix  }
+      | _                -> assert false) filtered
+end
+
+module Cm = ClauseMat
+
+(** {b Note} The compiling step creates a tree ready to be used for pattern
+    matching.  A tree guides the pattern matching by
+    - accepting constructors and filtering possible rules,
+    - guiding the matching in order to carry out as few atomic matchings as
+      possible by selecting the most appropriate term in the stack,
+    - storing terms from the stack that might be used in the right hand side,
+      because they match a pattern variable {!constructor:Patt} in the
+      {!field:lhs}.
+
+    The first bullet is ensured using {!val:specialize}, {!val:default} and
+    {!val:abstract} which allow to create new branches.
+
+    Efficiency is managed thanks to heuristics handled by the {!val:score}
+    function.
+
+    The last is managed by the {!val:env_builder} as follows.
+    The evaluation process uses, along with the tree, an array [vars] to store
+    terms matched against a pattern variable which is used in some
+    {!field:rhs}.  Each rule has an {!val:env_builder} mapping a index in the
+    [vars] array to a slot in the final environment (the slot [i] of a
+    [Patt(Some(i), _, _)]).  Note that the [vars] array can contain terms that
```

Comment:
Isn't [i] the index in the [vars] array and not the slot in the final environment? This is confusing.